install binary extensions

vilmibm opened the pull request:
This pull request teaches `gh extension install` how to detect, download, and install a precompiled ("binary") extension.

This is not full support for binary extensions; see #4194 for the remaining TODOs. However, this PR can be merged without breaking any existing functionality or advertising any new functionality.

I included tests.

Commit:
binary extension support in gh extension install

samcoe commented on pkg/cmd/extension/command.go:
```diff
@@ -102,15 +105,38 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
 					return err
 				}
 				if err := checkValidExtension(cmd.Root(), m, repo.RepoName()); err != nil {
+					// TODO i feel like this should check for a gh-foo script
```
I think rather than add that check here I would prefer just to rename `checkValidExtension` to `validExtensionName` or something similar to better reflect that it is only validating the name of the extension to be installed.

samcoe commented on pkg/cmd/extension/command.go:
```diff
@@ -102,15 +105,38 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					return m.InstallBin(client, repo)
+				}
+
+				hs, err := hasScript(client, repo)
```
What are your thoughts on moving this check above `isBinExtension`? Seems like we could then remove the `hasScript` check from `isBinExtension`. 

Or instead what are your thoughts on just encapsulating the type determination logic in its own function. Something like
```golang
type extensionType string

const (
	binary extensionType = "bin"
	git                  = "git"
)

func getExtensionType(client *http.Client, repo ghrepo.Interface) (extensionType, error) {
	hs, err := hasScript(client, repo)
	if err != nil {
		return "", err
	}
	if hs {
		return git, nil
	}
	
	_, err = fetchLatestRelease(client, repo)
 	if err != nil {
		return "", err
 	}
	return binary, nil
}
```

vilmibm replied:
I think my refactoring made this comment obsolete so I'm going to resolve but feel free to correct me~

samcoe commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +179,80 @@ func (m *Manager) InstallLocal(dir string) error {
 }
 
-func (m *Manager) Install(cloneURL string, stdout, stderr io.Writer) error {
+type BinManifest struct {
```
Seems like keeping the version in the manifest will be important for checking for updates.

samcoe commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +179,80 @@ func (m *Manager) InstallLocal(dir string) error {
+		return fmt.Errorf("failed to serialize manifest: %w", err)
+	}
+
+	manifestPath := filepath.Join(targetDir, "manifest.yml")
```
This is interesting. When we discussed having a manifest file for binary extensions I envisioned we would have a single manifest file for all binary extensions, maybe even including git extensions, that would be the source of truth for what is installed. This way we could read one file rather than navigating the filesystem. Thoughts on the trade off of that vs having a manifest file for each installed extension?

vilmibm replied:
I'd rather do a manifest per extension. I don't immediately see a lot to gain from having a centralized manifest file for all extensions and it technically means that we would need to implement file locking since it becomes a database of sorts with race conditions (unlikely but it would be irresponsible to not take into account).

samcoe commented on pkg/cmd/extension/manager_test.go:
```diff
@@ -190,18 +197,66 @@ func TestManager_Upgrade_NoExtensions(t *testing.T) {
 	assert.Equal(t, "", stderr.String())
 }
 
+func TestManager_InstallBin(t *testing.T) {
```
I think adding another test for when the extension does not have a binary for the given platform would be useful.

samcoe commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +179,80 @@ func (m *Manager) InstallLocal(dir string) error {
+
+	name := repo.RepoName()
+	targetDir := filepath.Join(m.installDir(), name)
+	// TODO clean this up if function errs?
```
I think this is important to address, mainly due to the fact that without doing any cleanup the subsequent install attempts will likely fail due to the weird state of the install folder. Or we could make sure that subsequent install attempt just overwrite the install folder and its contents if it exists?

vilmibm replied:
It's easy enough to call `gh extension remove foo` to clean up so I'm not sure how urgent this is...I'm not opposed to doing it but might save it for a future improvement.

mislav commented on pkg/extensions/extension.go:
```diff
@@ -16,7 +19,8 @@ type Extension interface {
 type ExtensionManager interface {
 	List(includeMetadata bool) []Extension
-	Install(url string, stdout, stderr io.Writer) error
+	InstallGit(url string, stdout, stderr io.Writer) error
+	InstallBin(client *http.Client, repo ghrepo.Interface) error
```
Do InstallGit and InstallBit have to be published? Couldn't they just stay an implementation detail of `Install`?

vilmibm replied:
I wasn't sure. Their signatures and implementations are so different it felt weird to combine them, but I'm certainly not opposed to that.

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,6 +246,26 @@ func checkValidExtension(rootCmd *cobra.Command, m extensions.ExtensionManager,
 }
 
+func isBinExtension(client *http.Client, repo ghrepo.Interface) (isBin bool, err error) {
+	hs, err := hasScript(client, repo)
+	if err != nil || hs {
+		return
```
So if the git repository happens to have a `gh-my-extension` file in its root, this will assume that it's a regular extension and not a binary extension and won't even look up releases?

Couldn't we just look up releases anyway and decide that it's a binary extension if it has binary attachments?

vilmibm replied:
we could i think. I had some reason for doing the script check that I don't think I recorded anywhere. I don't think I was thinking about strictly looking at the asset names. I'll switch to that and see how it feels.

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,6 +246,26 @@ func checkValidExtension(rootCmd *cobra.Command, m extensions.ExtensionManager,
+		return
+	}
+
+	isBin = true
```
Does this assume that the presence of any Releases in a repo means that it's a binary extension? Could we detect te presence of at least some assets looking like they are binary extensions and make an assumption based on that? I'm thinking that there might be non-binary extensions that want to utilize their Releases for purposes other than attaching a compiled binary to.

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +179,80 @@ func (m *Manager) InstallLocal(dir string) error {
+	suffix := m.platform()
+	var asset *releaseAsset
+	for _, a := range r.Assets {
+		if strings.HasSuffix(a.Name, suffix) {
```
Is the condition for a binary attachment only that it ends with `GOOS-GOARCH` as suffix? Seems straightforward enough 👍 But other than the suffix, is there a specific naming pattern for assets?

vilmibm replied:
I was only planning on enforcing the suffix of GOOS-GOARCH; I'm open to more restrictions but didn't immediately see value in it.

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +179,80 @@ func (m *Manager) InstallLocal(dir string) error {
+	}
+
+	if asset == nil {
+		return fmt.Errorf("%s unsupported for %s. Open an issue: `gh issue create -R%s/%s -t'Support %s'`",
```
Nit: even though a space after a shorthand flag isn't necessary, I think examples and printed instructions would be better if they included them for readability.

mislav reviewed (changes requested):
This is shaping up great. Love the simplicity of the approach!

vilmibm commented:
Thanks for the reviews y'all; I'm going to tweak the logic for detecting binary extensions and potentially refactor `InstallGit`/`InstallBin` then request re-review.

Commit:
more accurately check for binary extension

Commit:
nit

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
track installed tag name

Commit:
test for unsupported platform

vilmibm commented:
@mislav @samcoe I've:

- Changed the binary check to specifically look for release assets with platform names and not take into account the presence of a script
- Stored the release's tag in the manifest
- Refactored `InstallGit` and `InstallBin` into unexposed methods called by `Install`
- Added a test for unsupported binary platforms

Michael19841984 commented:
run

samcoe commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +180,104 @@ func (m *Manager) InstallLocal(dir string) error {
+		return err
+	}
+	if !hs {
+		// TODO open an issue hint, here?
```
I think just the error message is fine for now.

samcoe commented on pkg/cmd/extension/command.go:
```diff
@@ -220,9 +228,83 @@ func checkValidExtension(rootCmd *cobra.Command, m extensions.ExtensionManager,
 	return nil
 }
 
+func isBinExtension(client *http.Client, repo ghrepo.Interface) (isBin bool, err error) {
```
Since this is only used by the manager do you think the code should live in `manager.go`?

samcoe reviewed (approved):
Left one question. This looks great to me! Nice work!

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -104,13 +107,18 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
-				protocol, _ := cfg.Get(repo.RepoHost(), "git_protocol")
-				return m.Install(ghrepo.FormatRemoteURL(repo, protocol), io.Out, io.ErrOut)
+
+				return m.Install(client, repo, io, cfg)
```
Passing an HTTP client to the Extensions Manager is a leaky abstraction, IMO. The caller of `manager.Install()` shouldn't know of underlying implementation details of installing an extension, like API calls, git, etc.

If a specific HTTP API client needs to be used by Manager, that should be configured in the code that initializes the Manager https://github.com/cli/cli/blob/2c43436c7ab7dd0bfca11fee1b2d00d4c6c58a11/pkg/cmd/factory/default.go#L25-L25

vilmibm replied:
By this logic a manager should also be initialized with IOStreams and Config -- do you agree? I can make that refactor and it makes sense to me, I just noticed that in `trunk` it seems like setting an iostreams on a manager during its initialization was avoided.

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,9 +228,83 @@ func checkValidExtension(rootCmd *cobra.Command, m extensions.ExtensionManager,
 	return strings.TrimPrefix(n, "gh-")
 }
+
+func possibleDists() []string {
```
Could we just check for a single OS-ARCH combo instead of having to comb through this list? For example, every extension should publish a `linux-amd64` binary. Assuming that our official Action will also build this binary by default, we can assume that a Release that contains this binary is a binary extension.

vilmibm replied:
I didn't want to force authors to support a default OS-ARCH -- an extension might only make sense on windows, for example, because it integrates with the registry editor or something. What's your concern with the list?

mislav replied:
I guess the list is fine. It's just that the list feels long and it's not clear which combos of OS-ARCH made this cut in the first place and whether the list should be kept up to date with additional OS-ARCH combos in the future.

vilmibm replied:
it's the full list that go supports as output by `go tool dist list`. i figure new combinations are added so rarely that it wouldn't be an issue.

Commit:
refactor dependencies of ext manager

vilmibm commented:
@mislav I refactored extension manager to be initialized with dependencies -- sorry I didn't pick up on that pattern in the first place.

Commit:
linter appeasement

Commit:
restore cached client

mislav commented on pkg/extensions/extension.go:
```diff
@@ -16,7 +18,7 @@ type Extension interface {
 //go:generate moq -rm -out manager_mock.go . ExtensionManager
 type ExtensionManager interface {
 	List(includeMetadata bool) []Extension
-	Install(url string, stdout, stderr io.Writer) error
```
The fact that `Install()` now doesn't accept stdout/stderr and instead stores it internally via iostreams created a slight asymmetry with the signature of `Upgrade()`, which still takes stdout & stderr via arguments. I feel that it was not necessary for the addition of the binary extensions feature to change the logic around iostreams, and that stdout/stderr could have remained as arguments.

I do agree that passing in repo as `ghrepo.Interface` instead of a clone URL is better because it frees the Manager from having to parse out URLs internally to determine the repository 👍

vilmibm replied:
I was going to make the interfaces consistent in subsequent PRs (and have already updated `Upgrade` in my branch that's stacked on this one). Agreed that I didn't have to change the io handles, but as long as I was updating the Manager struct to avoid leaky abstractions I felt it was worth making things consistent.

mislav reviewed (approved):
This looks good. Thank you for the work!

heaths commented on pkg/cmd/extension/manager.go:
```diff
@@ -172,7 +192,104 @@ func (m *Manager) InstallLocal(dir string) error {
+	}
+	if !hs {
+		// TODO open an issue hint, here?
+		return errors.New("extension is uninstallable: missing executable")
```
@vilmibm FWIW, "uninstallable" more often means it can be uninstalled. More common in the space of acquisition that it's "not installable".

heaths commented on pkg/cmd/extension/manager.go:
```diff
@@ -346,3 +463,77 @@ func readPathFromFile(path string) (string, error) {
+	}
+
+	for _, a := range r.Assets {
+		dists := possibleDists()
```
Wouldn't it be more efficient to move this out of the `range` over assets? Depending on optimization, this is allocating at least a new slice with each iteration.

heaths commented:
@vilmibm is there documentation planned for the supported `{OS}-{ARCH}` values? I had to look in code to find them, and https://cli.github.com/manual/gh_extension doesn't look updated with this info yet.

Also, is there also discussion of the manifest.yml schema and how to use the file?

Edit: [seems](https://github.com/cli/cli/blob/cb89b8ca7a706cf51927477ad3f979ea4769e16e/pkg/cmd/extension/manager.go#L167-L195) it's used to point to a repo and specific tag, so is it expected that this file lives in a separate repo so not all source is pulled during a clone? (Could dig in further, but the point was asking for docs instead. 😀)
